src/Features/LanguageServer/Protocol/Features/EditAndContinue/EditAndContinueDiagnosticSource.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue;
+
+internal sealed class EditAndContinueDiagnosticSource
+{
+    public static async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(Document designTimeDocument, CancellationToken cancellationToken)
+    {
+        var designTimeSolution = designTimeDocument.Project.Solution;
+        var services = designTimeSolution.Services;
+
+        // avoid creating and synchronizing compile-time solution if Hot Reload/EnC session is not active
+        if (services.GetRequiredService<IEditAndContinueWorkspaceService>().SessionTracker is not { IsSessionActive: true } sessionStateTracker)
+        {
+            return [];
+        }
+
+        var applyDiagnostics = sessionStateTracker.ApplyChangesDiagnostics;
+
+        var compileTimeSolution = services.GetRequiredService<ICompileTimeSolutionProvider>().GetCompileTimeSolution(designTimeSolution);
+
+        var compileTimeDocument = await CompileTimeSolutionProvider.TryGetCompileTimeDocumentAsync(designTimeDocument, compileTimeSolution, cancellationToken).ConfigureAwait(false);
+        if (compileTimeDocument == null)
+        {
+            return applyDiagnostics;
```

Comment:
do these documents need to be filtered for the document being requested somehow?  e..g if you have two documents open it seems like right now we'd return the applydiagnostics twice

Reply:
These diagnostics are not document specific. How should we report them?

Reply:
Would adding a new diagnostic source that reports closed document and project diagnostics to `AbstractWorkspacePullDiagnosticsHandler.GetDiagnosticSourcesAsync` be appropriate?

Reply:
@tmat Hmm - I believe if you're reporting diagnostics through document pull - all the diagnostics will be reported against the document that was requested by the client, even if the diagnostic data points to some other document - see the textdoc identifier in the createreport call here - https://github.com/dotnet/roslyn/blob/6e2ac12e4c38780953fc41afb848e80c728075e5/src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs#L305

Closed document diagnostics are handled by workspace pull diagnostics, if its enabled.  I think in VSCode we have another alternative, which is to use the [related document diagnostics report](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#relatedFullDocumentDiagnosticReport), but that doesn't exist in VS IIRC.